common/persistence/nexus_incoming_service_manager.go

```diff
@@ -0,0 +1,159 @@
+		logger      log.Logger
+	}
+)
+
+func NewNexusIncomingServiceManager(
+	persistence NexusIncomingServiceStore,
+	serializer serialization.Serializer,
+	logger log.Logger,
+) NexusIncomingServiceManager {
+	return &nexusIncomingServiceManagerImpl{
+		persistence: persistence,
+		serializer:  serializer,
+		logger:      logger,
+	}
+}
+
+func (m *nexusIncomingServiceManagerImpl) GetName() string {
+	return m.persistence.GetName()
+}
+
+func (m *nexusIncomingServiceManagerImpl) Close() {
+	m.persistence.Close()
+}
+
+// GetNexusIncomingServicesTableVersion is a convenience method for getting the current nexus_incoming_services table
+// version by calling ListNexusIncomingServices with LastKnownTableVersion=0 and PageSize=0
+func (m *nexusIncomingServiceManagerImpl) GetNexusIncomingServicesTableVersion(ctx context.Context) (int64, error) {
+	tableVersion := int64(0)
+	resp, err := m.ListNexusIncomingServices(ctx, &ListNexusIncomingServicesRequest{
+		LastKnownTableVersion: 0,
+		PageSize:              0,
+	})
+	if resp != nil {
+		tableVersion = resp.TableVersion
+	}
+	return tableVersion, err
+}
+
+func (m *nexusIncomingServiceManagerImpl) ListNexusIncomingServices(
+	ctx context.Context,
+	request *ListNexusIncomingServicesRequest,
+) (*ListNexusIncomingServicesResponse, error) {
+	if request.PageSize < 0 {
+		return nil, ErrNonPositiveListNexusIncomingServicesPageSize
+	}
+
+	result := &ListNexusIncomingServicesResponse{}
+
+	resp, err := m.persistence.ListNexusIncomingServices(ctx, request)
+	if resp != nil {
+		result.TableVersion = resp.TableVersion
+	}
+	if err != nil {
+		return result, err
+	}
+
+	entries := make([]*persistencepb.NexusIncomingServiceEntry, len(resp.Services))
+	for i, entry := range resp.Services {
+		service, err := m.serializer.NexusIncomingServiceFromBlob(entry.Data)
+		if err != nil {
+			m.logger.Error(fmt.Sprintf("error deserializing nexus incoming service with ID:%v", entry.ServiceID), tag.Error(err))
+			return nil, err
+		}
+		entries[i].Id = entry.ServiceID
+		entries[i].Version = entry.Version
+		entries[i].Service = service
+	}
+
+	result.NextPageToken = resp.NextPageToken
+	result.Entries = entries
+	return result, nil
+}
+
+func (m *nexusIncomingServiceManagerImpl) CreateOrUpdateNexusIncomingService(
+	ctx context.Context,
+	request *CreateOrUpdateNexusIncomingServiceRequest,
+) (*CreateOrUpdateNexusIncomingServiceResponse, error) {
+	serviceBlob, err := m.serializer.NexusIncomingServiceToBlob(request.Entry.Service, enums.ENCODING_TYPE_PROTO3)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.persistence.CreateOrUpdateNexusIncomingService(ctx, &InternalCreateOrUpdateNexusIncomingServiceRequest{
+		LastKnownTableVersion: request.LastKnownTableVersion,
+		Service: InternalNexusIncomingService{
+			ServiceID: request.Entry.Id,
+			Version:   request.Entry.Version,
+			Data:      serviceBlob,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	request.Entry.Version++
```

Comment:
We're mutating the input here, that's not ideal and can lead to unexpected behavior.